learning_resources_search/tasks.py

```diff
@@ -108,6 +115,109 @@ def upsert_learning_resource(learning_resource_id):
+            if key == "department":
+                return LearningResourceDepartment.objects.get(department_id=val[0]).name
+            elif key == "offered_by":
+                return LearningResourceOfferor.objects.get(code=val[0]).name
+            return val[0]
+    return None
+
+
+def _infer_search_url(percolate_query):
+    """
+    Infer the search URL for the percolate query
+    """
+    original_query = OrderedDict(percolate_query.original_query)
+    query_string_params = {k: v for k, v in original_query.items() if v}
+    query_string = urlencode(query_string_params, doseq=True)
+    return f"{settings.SITE_BASE_URL}/search?{query_string}"
+
+
+def _group_percolated_rows(rows):
+    def key_func(x):
+        return (x["user_id"], x["group"])
+
+    grouped_data = {}
+    for key, group in itertools.groupby(rows, key_func):
+        context = list(group)
+        user_id = key[0]
+        group_name = key[1]
+        if key[0] not in grouped_data:
+            grouped_data[user_id] = {group_name: []}
+        if group_name not in grouped_data[user_id]:
+            grouped_data[user_id][group_name] = []
+        for ctx in context:
+            if ctx["user_id"] == user_id:
+                grouped_data[user_id][group_name].append(ctx)
+    return grouped_data
+
+
+def _get_percolated_rows(resources, subscription_type):
+    """
+    Get percolated rows for a list of learning resources and subscription type
+    """
+    rows = []
+    all_users = set()
+    # percolate each new learning resource to get matching queries
+    for resource in resources:
+        percolated = percolate_matches_for_document(resource.id).filter(
+            source_type=subscription_type
+        )
+        if percolated.count() > 0:
+            percolated_users = set(percolated.values_list("users", flat=True))
+            all_users.update(percolated_users)
+            query = percolated.first()
+            rows.extend(
+                [
+                    {
+                        "resource_url": resource.url,
+                        "resource_title": resource.title,
+                        "user_id": user,
+                        "group": _infer_percolate_group(query),
+                        "search_url": _infer_search_url(query),
+                    }
+                    for user in percolated_users
+                ]
+            )
+
+    return rows
+
+
+@app.task(bind=True)
+def send_subscription_emails(self, subscription_type, period="daily"):
+    """
+    Send subscription emails by percolating matched documents
+    """
+    delta = datetime.timedelta(days=1)
+    if period == "weekly":
+        delta = datetime.timedelta(days=7)
+    since = now_in_utc() - delta
+    new_learning_resources = LearningResource.objects.filter(
+        published=True, created_on__gt=since
+    )
+    rows = _get_percolated_rows(new_learning_resources, subscription_type)
+    template_data = _group_percolated_rows(rows)
```

Comment:
The size of this, in terms of memory, might eventually become a concern too if/when we have 100K+ users, depending on how many subscribe to searches and how many matches are returned.  But we can worry about that for another day.